### PR TITLE
Frontend api devices

### PR DESCRIPTION
**Pull request by thisconnect**
- adds api/devices and api/devicessync
- rewrite app.tsx to use api/devices*
- change get started and unlock routing to directly go to account-summary
- updated app / device routing

### CODE REVIEW

**Comment by thisconnect** [comment]
This is my little testing routine

```
- tap side
- unlock
- should see account summary

- unplug/replug

- tap side
- unplug/replug
- should see waiting screen

- tap side
- restart app
- should see unlock screen

- factory reset
- tap side
- should see verify pairing code
- unplug/replug
- tap side
- should see verify pairing code
- now either create or restore


```

**Comment by thisconnect** [comment]
Github shows 5 commits but it is only the 2 latest, the other 3 are now merged (ff) into frontend-new-api

**Comment by thisconnect** [comment]
> Please rebase; I was halfway through reviewing when I noticed that some of the commits were already merged 🙈

Sorry about that, but I left a comment. Not sure why it showed the 3 merged commits, not it looks ok only 2 commits.

https://github.com/digitalbitbox/bitbox-wallet-app/pull/1194#issuecomment-791506273

File: frontends/web/src/api/devices.ts [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+type TSuccess = {
+    success: true;
+}
+
+// if the backend uses maybeBB02Err
+type TFail = {
+    code?: number;
+    message?: string;
+    success: false;
```
Comment (benma) [inline frontends/web/src/api/devices.ts:38]: success: boolean?

File: frontends/web/src/api/devices.ts [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    | 'unpaired' | 'pairingFailed'
+    | 'uninitialized' | 'seeded' | 'initialized'
+    | 'require_firmware_upgrade' | 'require_app_upgrade';
+
+export type TStatusBB01 = 'unpaired' | 'pairingFailed'
+    | 'uninitialized' | 'seeded' | 'initialized'
+    | 'require_firmware_upgrade' | 'require_app_upgrade';
+
+type TSuccess = {
+    success: true;
```
Comment (benma) [inline frontends/web/src/api/devices.ts:31]: boolean?

File: frontends/web/src/api/devices.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+// firmware
+
+export type TBundledFirmwareVersion = {
+    canUpgrade: boolean;
+    currentVersion: string;
+    newVersion: string;
+}
+
+export const getBundledFirmwareVersion = (
```
Comment (benma) [inline frontends/web/src/api/devices.ts:135]: many of these funcs in here seem unused. please remove them until they are needed, otherwise it's difficult to see if they are correct without being able to test their use.

File: frontends/web/src/api/devices.ts [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+import { apiGet, apiPost } from '../utils/request';
+
+type TProuctName = 'bitbox' | 'bitbox02' | 'bitbox02-bootloader';
+
+export type TStatus = 'connected'
+    | 'unpaired' | 'pairingFailed'
+    | 'uninitialized' | 'seeded' | 'initialized'
+    | 'require_firmware_upgrade' | 'require_app_upgrade';
+
+export type TStatusBB01 = 'unpaired' | 'pairingFailed'
```
Comment (benma) [inline frontends/web/src/api/devices.ts:26]: can you make a bitbox02.ts and a bitbox01.ts? it would be a lot clearer imho.

**Review by benma — CHANGES_REQUESTED** [review]

File: frontends/web/src/api/devices.ts (hunk 2 above)
Comment (thisconnect) [reply frontends/web/src/api/devices.ts:31]: Similar to TFail, TSuccess is always success true

File: frontends/web/src/api/devices.ts (hunk 1 above)
Comment (benma) [reply frontends/web/src/api/devices.ts:38]: I see!

Comment (thisconnect) [reply frontends/web/src/api/devices.ts:38]: With that TypeScript knows which other properties the object has.

File: frontends/web/src/api/devices.ts (hunk 3 above)
Comment (thisconnect) [reply frontends/web/src/api/devices.ts:135]: This time, I didn't copy from the existing code, but typed the api by reading go backend, so I need to back it up :)

FYI on this branch there is still unused `api/account.ts`, we should merge master into this branch.

File: frontends/web/src/api/devices.ts (hunk 4 above)
Comment (thisconnect) [reply frontends/web/src/api/devices.ts:26]: I like, but there is shared functiontions and types that are not specific to bitbox01 or bitbox02, should we keep the shared ones in `api/devices.ts` or have it in `api/bitbox02.ts`?

File: frontends/web/src/api/devices.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { apiGet } from '../utils/request';
+
+export type TProuctName = 'bitbox' | 'bitbox02' | 'bitbox02-bootloader';
```
Comment (benma) [inline frontends/web/src/api/devices.ts:19]: typo. also does not need to be expoerted?

File: frontends/web/src/api/devices.ts (hunk 4 above)
Comment (benma) [reply frontends/web/src/api/devices.ts]: concept ack, nice!

File: frontends/web/src/api/devices.ts (hunk 5 above)
Comment (thisconnect) [reply frontends/web/src/api/devices.ts]: removed `export`

File: frontends/web/src/api/devices.ts (hunk 4 above)
Comment (thisconnect) [reply frontends/web/src/api/devices.ts]: ok updated and since `api/bitbox01.ts` and `api/bitbox02.ts` isn't used it is not anymore part of this PR

**Comment by thisconnect** [comment]
@benma squashed and removed unused code, actually didn't even add `bitbox01.ts` and `bitbox02.ts` PTAL

**Comment by thisconnect** [comment]
There is currently a routing issue after unlock.

**Comment by thisconnect** [comment]
Because this PR changes how the app component gets devices, before via decorator props and now directly into state added 1 commit to fix and simplify the routing on new devices/accounts.

Tests:
* plugin device: route to devices
* unlock device: route to account-summary (default view still quickly flickers)
* unplug device: route and show default view (from account-summary, account, send, receive, buy, settings, manage-backup)
* unplug device: show default view (from device, doesn't route)
* restart app with unlocked bb: restart servewallet, navigate to / routes to account-summary 
* devmode: reload should stay on the same view (account-summary, account, send, receive, buy, settings)
    - Bug: reloading manage-backup redirects to account-summary
* devmode: restart app, similar to restart app but already on a route, behaves same as devmode just reload with the manage-backup bug

Please let me know if you can think of any other situations that have to be tested.

File: frontends/web/src/app.tsx [hunk 6]
```diff
@@ -116,14 +113,33 @@ class App extends Component<Props, State> {
...
-            .then(accounts => this.setState({ accounts }, this.maybeRoute))
-            .catch(console.error);
+        getDeviceList()
+        .then(devices => {
+            return getAccounts()
+            .then(accounts => {
+                this.setState({ accounts, devices }, this.maybeRoute);
+            });
+        })
+        .catch(console.error);
```
Comment (benma) [inline frontends/web/src/app.tsx:124]: it is not so nice that we now have to both manually call the endpoints and also subscribe the endpoints, each time doing similar/the same things after it finishes.

the previous subscribe decorator took care of both. can your new subscribe api also take care of both at the same time?

food for thought, not in this PR.

File: frontends/web/src/app.tsx [hunk 7]
```diff
@@ -116,14 +113,33 @@ class App extends Component<Props, State> {
...
 
         this.unsubscribeList.push(
             syncAccountsList(accounts => {
                 this.setState({ accounts }, this.maybeRoute);
             }),
+            syncDeviceList(devices => {
+                const oldDeviceIDList = Object.keys(this.state.devices);
```
Comment (benma) [inline frontends/web/src/app.tsx:130]: Scattering routing logic makes it harder to reason about the routing conditions and rules.

Why not simply call `this.maybeRoute` like above with the accounts, and keep the routing code there?

**Review by benma — APPROVED** [review]

File: frontends/web/src/app.tsx (hunk 6 above)
Comment (thisconnect) [reply frontends/web/src/app.tsx:124]: Currently:
- App.tsx and devices.jsx (old BB01) are the only places where we subscribe and fetch `devices/registered`.
- Also App.tsx is the only place where it subscribes to syncAccountsList, which are then passed as props to the children.

> each time doing similar/the same things after it finishes

Was doing the same/similar thing, but not anymore with this PR. The routing fix in this PR needs to compare the current devices list with the new one, so it can't be in mayberoute.

Personally I don't like mayberoute it was the cause of various routing bugs in the past and has to maintain a centralized list of exception routing for different views, it is especially hard to follow if there is additional routing in views.

I'd prefer if mayberoute could go away in the future and each view decides for itself if it should be routed to `/` when the device is unplugged or not.
i.e. something like `import { routeToDefaultViewOnDeviceUnplug } from 'utils/routingHelper'` and  `this.unsubscribe = routeToDefaultViewOnDeviceUnplug();`

The device subscription probably makes sense that it stays in app.tsx but the accountlist fetching + subscription could be in a reusable component and only included in views that actually need it.

File: frontends/web/src/app.tsx (hunk 7 above)
Comment (thisconnect) [reply frontends/web/src/app.tsx:130]: without this PR the app routes many times back and forth and after routing happen routes again and was really hard to follow.

This PR reduces the amount of routing and depends on checking if there is a new device, this can't be done in  maybeRoute.

Also there are other arguments against maybeRoute see my other comment.

**Comment by thisconnect** [comment]
> Please let me know if you can think of any other situations that have to be tested.

@benma I'll try to test with BB01, but please let me know if you can think of any other testing that I forgot.

File: frontends/web/src/app.tsx (hunk 6 above)
Comment (benma) [reply frontends/web/src/app.tsx:124]: > Personally I don't like mayberoute it was the cause of various routing bugs in the past and has to maintain a centralized list of exception routing for different views, it is especially hard to follow if there is additional routing in views.

Fyi, before we had maybeRoute, we had `route()` scattered and it also had many routing bugs and routing regressions. I had put in one place hoping that it would improve that.

But my comment here was actually not about maybeRoute() at all. Just an observation that for one `subscribe` decorator now, there are two very similar calls that have to be done now, e.g. `getDeviceList` and `syncDeviceList`. Not a problem for now.

File: frontends/web/src/app.tsx [hunk 8]
```diff
@@ -116,14 +113,33 @@ class App extends Component<Props, State> {
             }
         });
 
-        getAccounts()
-            .then(accounts => this.setState({ accounts }, this.maybeRoute))
-            .catch(console.error);
+        getDeviceList()
```
Comment (benma) [inline frontends/web/src/app.tsx:116]: A comment here explaining why exactly the two calls are chained like this will help in the future, I think.

File: frontends/web/src/app.tsx (hunk 6 above)
Comment (thisconnect) [reply frontends/web/src/app.tsx:124]: > Fyi, before we had maybeRoute, we had route() scattered and it also had many routing bugs and routing regressions. I had put in one place hoping that it would improve that.

👍 

For now, I think we should stick to the current way, devices routing done by app with a list of exceptions. 

Alternative would be to not route in app.tsx at all and make each view take care about the logic (this could be done with reusable modules to be DRY).

File: frontends/web/src/app.tsx (hunk 8 above)
Comment (thisconnect) [reply frontends/web/src/app.tsx]: You're right, the requests can be done in parallel, no reason to wait, changed back and squashed 
https://github.com/digitalbitbox/bitbox-wallet-app/compare/d27e84efd912e3a1814cf96d168e64aa45ed3323..1711d06eac771a604f93ac9d2f42085f39fcb534

Also tested BB01 pairing, sending, restore form backup and quickly tested BB02 one more time, will merge now.
